Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64442

Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

peterbell10 committed:
Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30735341).

ngimel commented:
Build errors are real

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/MultinomialKernel.cu:
```diff
@@ -71,7 +67,7 @@ void renormRows(Tensor& t) {
   AT_DISPATCH_FLOATING_TYPES_AND_HALF(t.scalar_type(), "renormRows_cuda", [&] {
     renormRowsL1<scalar_t>
-        <<<grid, block, block.x * sizeof(scalar_t),
+        <<<grid, block, (block.x / C10_WARP_SIZE) * sizeof(scalar_t),
```
BlockReduceSum is calling `warp_shfl_down` with full mask, which requires all threads to participate. So you need to ensure that block.x is a multiple of WARP_SIZE (and make sure that added threads don't attempt to read values)

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30735341).

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cu:
```diff
@@ -182,16 +183,19 @@ void fused_mode(
     case 8:
     case 4:
-    case 2:
-      handle_fused_mode<32, scalar_t>(
```
32 seems like a weird choice to me. Why have a dedicated kernel with fewer threads than there are in a single warp?

ngimel replied:
Yeah, that's a weird choice indeed, probably they missed that `size/2`

peterbell10 commented on aten/src/ATen/native/cuda/MultinomialKernel.cu:
```diff
@@ -60,10 +64,10 @@ void renormRows(Tensor& t) {
 
   dim3 grid(rows < numSM * 4 ? rows : numSM * 4);
-  dim3 block(cols < maxThreads ? cols : maxThreads);
+  dim3 block(std::min(maxThreads, C10_WARP_SIZE * div_up(cols, C10_WARP_SIZE)));
```
Rounding up to a multiple of `C10_WARP_SIZE` is okay because the kernel structure is:
```cpp
    for (int64_t col = threadIdx.x; col < cols; col += blockDim.x) {
      // local sum
    }

    sum = cuda_utils::BlockReduceSum(sum, smem);
    // broadcast block sum to all threads...

    for (int64_t col = threadIdx.x; col < cols; col += blockDim.x) {
        // use block sum
    }
```

peterbell10 commented on aten/src/ATen/native/cuda/Embedding.cu:
```diff
@@ -320,7 +316,7 @@ Tensor & embedding_renorm_cuda_(Tensor & self, const Tensor & indices,
     AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, self.scalar_type(), "embedding_backward", [&] {
       using accscalar_t = acc_type<scalar_t, true>;
-      renorm_kernel<<<grid, block, 128 * sizeof(accscalar_t), stream>>>(
+      renorm_kernel<<<grid, block, (block.x / C10_WARP_SIZE) * sizeof(accscalar_t), stream>>>(
```
Block size here is a constant 128 so is always a multiple of `C10_WARP_SIZE`.

ngimel replied:
Yep, to be on the safe side, can you please contexpr `block` and put in a `static_assert` to verify? Remember, some builds require 2 arg version of static_assert, I'm constantly bitten by this.

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

ngimel commented:
ROCm failure looks real (it's in test_mode)

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cuda/TensorModeKernel.cuh:
```diff
@@ -374,43 +383,43 @@ __global__ void compute_mode(
-  // go back to using the (index, flag) buffer arrangement. First, we mark
-  // indices that are equal to the mode, i.e B[i] = true if input[i] == mode,
-  // and initialize C[i] to be the index
+  // Finally, we need to find "an" index of the mode in the input
+  // Tensor. The API does not constrain which index we pick, but here
+  // we always pick the largest index. We store the index if the value
+  // is the mode, or 0 otherwise. Then find the maximum value.
```
This fixes the ROCm failures. The old code happened to always pick the largest index due to the way the reductuion was structured. However, the warp shuffle based reduction has a different order of comparisons, so didn't have the same end result. This operator actually guarantees the largest index is returned, regardless of the reduction order.

ngimel replied:
Out of curiosity, doesn't WARP_SHFL_DOWN also always the last (largest) index in the end?

peterbell10 replied:
`WarpReduce` goes from biggest offset to smallest offset. So, if priority goes to the right hand side then element 7 has higher priority than 8 or 9 because 8 = 0 mod 8 and 9 = 1 mod 8 are to the left of 7 = 7 mod 8.

peterbell10 commented on aten/src/ATen/native/cuda/block_reduce.cuh:
```diff
@@ -10,6 +10,7 @@ namespace native {
 namespace cuda_utils {
 
 constexpr int kCUDABlockReduceNumThreads = 512;
+constexpr int kCUDABlockReduceMaxThreads = C10_WARP_SIZE * C10_WARP_SIZE;
```
Can you explain your logic? The function does two reductions, each of which divides the width by `C10_WARP_SIZE`. So the max size should be `C10_WARP_SIZE` squared.

peterbell10 replied:
Oh, you mean that the device `maxThreadsPerBlock` is likely to be 1024? I meant this to reflect the limitation of the algorithm, not hardware. If you call with more threads than this value, the algorithm silently fails.

ngimel replied:
Yeah, in terms of algorithm limitations you are right, both on ROCm and on CUDA number of threads per block cannot be larger anyway (on ROCm it can only be smaller, as this evaluates to 4096, and maxThreadsPerBlock is 1024( , but it's ok to leave it as is. Can you please leave a comment that it's an algorithm limitation?

peterbell10 committed:
Update on "Migrate uses of THCReduceApplyUtils to cuda_utils::BlockReduce"


Differential Revision: [D30735341](https://our.internmc.facebook.com/intern/diff/D30735341)

[ghstack-poisoned]

ngimel reviewed (approved):
Thanks!

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30735341).

ngimel commented:
Sorry, have to revert as it breaks windows build:
```
2021-09-08T19:21:38.0093861Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): error C2131: expression did not evaluate to a constant
2021-09-08T19:21:38.0096133Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): note: failure was caused by a read of a variable outside its lifetime
2021-09-08T19:21:38.0098239Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): note: see usage of 'this'
2021-09-08T19:21:38.0100332Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): error C2131: expression did not evaluate to a constant
2021-09-08T19:21:38.0102646Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): note: failure was caused by a read of a variable outside its lifetime
2021-09-08T19:21:38.0104738Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1214406120\aten\src\ATen\native\cuda\Embedding.cu(299): note: see usage of 'this'
```
Probably num_threads assignment and static_assert can be moved outside of dispatch macro?
